src/realm/sync/config.hpp

```diff
@@ -165,6 +165,10 @@ struct SyncConfig {
     std::function<void(std::shared_ptr<Realm> before_frozen, std::shared_ptr<Realm> after, bool did_recover)>
         notify_after_client_reset;
 
+    // Will be called after a download message is received and validated by the client but befefore it's been
+    // transformed or applied. To be used in testing only.
+    std::function<void(std::weak_ptr<SyncSession>)> on_download_message_received_hook;
```

Comment:
do you really need the SyncSession here? It seems to be needed only by tests. Can you use `realm->sync_session()` instead?